examples/adwords_axis/src/main/java/adwords/axis/v201809/migration/MigrateToExtensionSettings.java

```diff
@@ -0,0 +1,594 @@
+              siteLinkFromFeed.line2 = attributeValue.getStringValue();
+              break;
+            case PLACEHOLDER_FIELD_LINE_3_TEXT:
+              siteLinkFromFeed.line3 = attributeValue.getStringValue();
+              break;
+            default:
+              // Ignore attributes that do not map to a predefined placeholder field.
+              break;
+          }
+        }
+      }
+
+      feedItems.put(feedItem.getFeedItemId(), siteLinkFromFeed);
+    }
+
+    return feedItems;
+  }
+
+  /**
+   * Gets the feed mapping for a feed.
+   *
+   * @return a multimap from feed attribute ID to the set of field IDs mapped to the attribute
+   */
+  private static Multimap<Long, Integer> getFeedMapping(AdWordsServicesInterface adWordsServices,
+      AdWordsSession session, Feed feed, long placeholderType) throws RemoteException {
+    // Get the FeedMappingService.
+    FeedMappingServiceInterface feedMappingService =
+        adWordsServices.get(session, FeedMappingServiceInterface.class);
+    String query = String.format(
+        "SELECT FeedMappingId, AttributeFieldMappings WHERE FeedId = %d and PlaceholderType = %d "
+        + "AND Status = 'ENABLED'", feed.getId(), placeholderType);
+
+    Multimap<Long, Integer> attributeMappings = HashMultimap.create();
+    int offset = 0;
+    FeedMappingPage feedMappingPage;
+
+    do {
+      String pageQuery = String.format(query + " LIMIT %d, %d", offset, PAGE_SIZE);
+      feedMappingPage = feedMappingService.query(pageQuery);
+      if (feedMappingPage.getEntries() != null) {
+        // Normally, a feed attribute is mapped only to one field. However, you may map it to more
+        // than one field if needed.
+        for (FeedMapping feedMapping : feedMappingPage.getEntries()) {
+          for (AttributeFieldMapping attributeMapping : feedMapping.getAttributeFieldMappings()) {
+            attributeMappings.put(attributeMapping.getFeedAttributeId(),
+                attributeMapping.getFieldId());
+          }
+        }
+      }
+      offset += PAGE_SIZE;
+    } while (offset < feedMappingPage.getTotalNumEntries());
+
+    return attributeMappings;
+  }
+
+  /** Returns a list of all enabled feeds. */
+  private static List<Feed> getFeeds(
+      AdWordsServicesInterface adWordsServices, AdWordsSession session) throws RemoteException {
+    FeedServiceInterface feedService = adWordsServices.get(session, FeedServiceInterface.class);
+    String query = "SELECT Id, Name, Attributes WHERE Origin = 'USER' AND FeedStatus = 'ENABLED'";
+
+    List<Feed> feeds = new ArrayList<>();
+    int offset = 0;
+    FeedPage feedPage;
+
+    do {
+      String pageQuery = String.format(query + " LIMIT %d, %d", offset, PAGE_SIZE);
+      feedPage = feedService.query(pageQuery);
+      if (feedPage.getEntries() != null) {
+        feeds.addAll(Arrays.asList(feedPage.getEntries()));
+      }
+      offset += PAGE_SIZE;
+    } while (offset < feedPage.getTotalNumEntries());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 323:330, 351:358, 570:577.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.